dipy/align/tests/test_mattes.py

```diff
@@ -0,0 +1,776 @@
+        metric = MattesBase(32)
+        metric.setup(static, moving, smask, mmask)
+
+        # 1. Update the joint distribution
+        sp_to_moving = np.linalg.inv(moving_aff)
+        samples_moving_grid = (sp_to_moving.dot(samples.T).T)[...,:dim]
+        intensities_moving, inside = interp_method(moving.astype(np.float32),
+                                                   samples_moving_grid)
+        intensities_moving = np.array(intensities_moving, dtype=np.float64)
+        metric.update_pdfs_sparse(intensities_static, intensities_moving)
+        J0 = np.copy(metric.joint)
+
+        # 2. Update the joint distribution gradient (the derivative of each
+        # histogram cell w.r.t. the transform parameters). This requires
+        # among other things, the spatial gradient of the moving image.
+        theta = transform.get_identity_parameters().copy()
+        grid_to_space = np.eye(dim + 1)
+        spacing = np.ones(dim, dtype=np.float64)
+        shape = np.array(static.shape, dtype=np.int32)
+        mgrad, inside = vf.gradient(moving.astype(np.float32), moving_aff,
+                                    spacing, shape, grid_to_space)
+        metric.update_gradient_dense(theta, transform, static.astype(np.float64),
+                                     moving.astype(np.float64), grid_to_space,
+                                     mgrad, smask, mmask)
+
+        # 3. Update the metric (in this case, the Mutual Information) and its
+        # gradient, which is computed from the joint density and its gradient
+        metric.update_mi_metric(update_gradient=True)
+
+        # Now we can extract the value and gradient of the metric
+        # This is the gradient according to the implementation under test
+        val0 = metric.metric_val
+        actual = np.copy(metric.metric_grad)
+
+        # Compute the gradient using finite-diferences
+        n = transform.get_number_of_parameters()
+        expected = np.empty_like(actual)
+        for i in range(n):
+            dtheta = theta.copy()
+            dtheta[i] += h
+
+            T = transform.param_to_matrix(dtheta)
+            shape = np.array(static.shape, dtype=np.int32)
+            sp_to_moving = np.linalg.inv(moving_aff).dot(T)
+            samples_moving_grid = (sp_to_moving.dot(samples.T).T)[...,:dim]
+            intensities_moving, inside = interp_method(moving.astype(np.float32),
+                                                       samples_moving_grid)
+            intensities_moving = np.array(intensities_moving, dtype=np.float64)
+            metric.update_pdfs_sparse(intensities_static, intensities_moving)
+            metric.update_mi_metric(update_gradient=False)
+            val1 = metric.metric_val
+            expected[i] = (val1 - val0) / h
+
+        dp = expected.dot(actual)
+        enorm = np.linalg.norm(expected)
+        anorm = np.linalg.norm(actual)
+        nprod = dp / (enorm * anorm)
+        assert_equal(nprod > 0.99, True)
+
```

Comment:
I fear there is some considerable replication between the dense and sparse versions of the tests. Consider having the duplicates as singletons in different functions or merge the dense and sparse in the same test.